core/Pos/Core/Genesis/Generate.hs

```diff
@@ -2,128 +2,149 @@
+    -- ^ Non-avvm balances
+    , ggdAvvm             :: !GenesisAvvmBalances
+    -- ^ Avvm balances
     , ggdBootStakeholders :: !GenesisWStakeholders
     -- ^ Set of boot stakeholders (richmen addresses or custom addresses)
-    , ggdGtData           :: !VssCertificatesMap
+    , ggdVssCerts         :: !GenesisVssCertificatesMap
     -- ^ Genesis vss data (vss certs of richmen)
-    , ggdSecretKeys       :: !(Maybe [(SecretKey, EncryptedSecretKey, VssKeyPair)])
+    , ggdSecrets          :: !(Maybe GeneratedSecrets)
+    }
+
+data GeneratedSecrets = GeneratedSecrets
+    { gsSecretKeys    :: ![(SecretKey, EncryptedSecretKey, VssKeyPair)]
     -- ^ Secret keys for non avvm addresses
-    , ggdFakeAvvmSeeds    :: !(Maybe [ByteString])
+    , gsFakeAvvmSeeds :: ![ByteString]
     -- ^ Fake avvm seeds (needed only for testnet)
     }
 
-generateGenesisData :: GenesisInitializer -> GeneratedGenesisData
-generateGenesisData TestnetInitializer{..} = deterministic (serialize' tiSeed) $ do
-    (fakeAvvmDistr, seeds) <- generateFakeAvvmGenesis tiFakeAvvmBalance
-    testnetGenData <- generateTestnetData tiTestBalance tiDistribution
-    let testnetDistr = ggdNonAvvmDistr testnetGenData
-    pure $
-        testnetGenData
-        { ggdNonAvvmDistr = testnetDistr ++ fakeAvvmDistr
-        , ggdFakeAvvmSeeds = Just seeds
-        }
-generateGenesisData MainnetInitializer{..} =
-    GeneratedGenesisData [] miBootStakeholders miVssCerts Nothing Nothing
-
--- | Generates keys and vss certs for testnet data.
-generateTestnetData
-    :: (MonadRandom m)
-    => TestnetBalanceOptions
-    -> TestnetDistribution
-    -> m GeneratedGenesisData
-generateTestnetData tso@TestnetBalanceOptions{..} distrSpec = do
+generateGenesisData
+    :: (HasProtocolConstants, HasGenesisBlockVersionData)
+    => GenesisInitializer
+    -> Word64
+    -> GeneratedGenesisData
+generateGenesisData (TestnetInitializer{..}) maxTnBalance = deterministic (serialize' tiSeed) $ do
+    let TestnetBalanceOptions{..} = tiTestBalance
+    (fakeAvvmDistr, seeds, fakeAvvmBalance) <- generateFakeAvvmGenesis tiFakeAvvmBalance
     (richmenList, poorsList) <-
-        (,) <$> replicateM (fromIntegral tboRichmen) (generateSecretsAndAddress Nothing tboUseHDAddresses)
-            <*> replicateM (fromIntegral tboPoors)   (generateSecretsAndAddress Nothing tboUseHDAddresses)
-
-    let skVssCerts = map (\(sk, _, _, vc, _) -> (sk, vc)) $ richmenList ++ poorsList
-    let richSkVssCerts = take (fromIntegral tboRichmen) skVssCerts
-    let secretKeys = map (\(sk, hdwSk, vssSk, _, _) -> (sk, hdwSk, vssSk)) $ richmenList ++ poorsList
+        (,) <$> replicateM (fromIntegral tboRichmen)
+                           (generateSecretsAndAddress Nothing tboUseHDAddresses)
+            <*> replicateM (fromIntegral tboPoors)
+                           (generateSecretsAndAddress Nothing tboUseHDAddresses)
+
+    let richSkVssCerts = map (\(sk, _, _, vc, _) -> (sk, vc)) $ richmenList
```

Comment:
VSS certs for poor men not needed ever? I guess the assumption is that a generated genesis never gives any stake to poor men?